### PR TITLE
Add slot bounds check and unit tests for pluginsd_parser

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary
 - Fixes parent-agent crashes caused by bad SLOT: values in streaming/pluginsd data.
  - A misbehaving or malicious child/plugin could send a huge slot number, making the parent try to grab a massive chunk of memory (crash) or write past the end of an array (corruption).
  This PR puts sensible upper limits on those values.
  - Oversized slots are now safely ignored (with a rate-limited warning) and the data still gets processed normally via the regular lookup path — nothing is dropped.
  - Separate, generous limits for charts (1,000,000) and dimensions (65,535), so real-world setups are never affected.
  - Added tests that confirm valid slots still work and out-of-range values are rejected.

---
## Summary by cubic
Adds slot bounds checking in the pluginsd parser to prevent crashes from oversized SLOT values, plus a guard against array allocation overflows. Clarifies fallback behavior when slots are invalid so charts and dimensions keep working without data loss.

- **Bug Fixes**
  - Enforce per-command max slot caps (charts 1,000,000; dimensions 65,535) across begin/chart/dimension/set and replication paths.
  - Validate `max_slot` in `pluginsd_parse_rrd_slot` (must fit `ssize_t`) and treat over-limit values as uncached with a rate-limited warning.
  - Clarify `slot < 1` handling: charts fall back to id lookup; dimensions use the non-slotted path.
  - Add allocation overflow guard in `prd_array_create`.

- **Tests**
  - Add slot-bounds unit tests covering valid, boundary, malformed/negative/hex, and overflow cases for both chart and dimension limits.
  - Add guard for `max_slot >= 1` to prevent underflow in boundary tests and simplify `snprintfz` usage.

<sup>Written for commit 9f9a1782de408f8b92dd35198ac1847088275af1. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/netdata/netdata/pull/22598?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/plugins.d/pluginsd_internals.h [hunk 1]
```diff
@@ -369,13 +369,21 @@ static inline RRDSET *pluginsd_find_chart(RRDHOST *host, const char *chart, cons
...
+        unsigned long long parsed_slot = str2ull_encoded(&id[5]);
+        if(unlikely(parsed_slot > max_slot)) {
+            nd_log_limit_static_global_var(erl_slot, 1, 0);
+            nd_log_limit(&erl_slot, NDLS_COLLECTORS, NDLP_WARNING,
+                         "PLUGINSD: ignoring invalid SLOT value '%s' above the supported maximum %zu",
+                         &id[5], max_slot);
+            slot = 0;
+        }
+        else
+            slot = (ssize_t)parsed_slot;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/plugins.d/pluginsd_internals.h:401]: False positive. The reads are bounded by short-circuit evaluation plus NUL termination: && evaluates left-to-right, so id[k] is only read after id[0..k-1] matched the non-NUL chars 'S','L','O','T'. For id[4] to be read, the word must already hold those 4 non-NUL bytes, so the buffer is at least "SLOT\0" (5 bytes) and id[4] is at worst the NUL terminator -- always in bounds. The "BEGIN X" example reads only id[0]='X', mismatches 'S', and stops. The token also comes from get_word() which returns NUL-terminated words. Added a source comment documenting this. Note this prefix check is pre-existing (commit da32dd8), not introduced by this PR.

File: src/plugins.d/pluginsd_internals.h [hunk 2]
```diff
@@ -369,13 +369,28 @@ static inline RRDSET *pluginsd_find_chart(RRDHOST *host, const char *chart, cons
...
+            nd_log_limit(&erl_slot, NDLS_COLLECTORS, NDLP_WARNING,
+                         "PLUGINSD: ignoring invalid SLOT value '%s' above the supported maximum %zu",
+                         &id[5], max_slot);
+            // slot 0 means: the SLOT word was present (so the caller still advances its
+            // word index), but the value is unusable as a cache index. Downstream cache
+            // paths treat slot < 1 as "uncached" and fall back to lookup by id.
+            slot = 0;
+        }
+        else
+            slot = (ssize_t)parsed_slot;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/plugins.d/pluginsd_internals.h:401]: Fair API-robustness point, though not currently reachable: both callers pass small compile-time caps (PLUGINSD_CHART_SLOT_MAX=1000000, PLUGINSD_DIMENSION_SLOT_MAX=65535), each far below SSIZE_MAX on 32-bit and 64-bit, so values in (SSIZE_MAX, max_slot] cannot occur. To keep the API self-consistent against a future caller, I added an internal_fatal(max_slot > (size_t)SSIZE_MAX, ...) guard at the top of pluginsd_parse_rrd_slot(). Since the function is ALWAYS_INLINE and max_slot is a constant at every call site, the check folds away in production builds and only documents/enforces the contract in debug builds.

File: src/plugins.d/pluginsd_parser.c [hunk 3]
```diff
@@ -1513,7 +1513,65 @@ void parser_init_repertoire(PARSER *parser, PARSER_REPERTOIRE repertoire) {
...
+              PLUGINSD_KEYWORD_SLOT ":%zu", max_slot - 1);
+    cases[n - 3].expected = (ssize_t)(max_slot - 1);
+
+    snprintfz(cases[n - 2].slot_word, sizeof(cases[n - 2].slot_word) - 1,
+              PLUGINSD_KEYWORD_SLOT ":%zu", max_slot);
+    cases[n - 2].expected = (ssize_t)max_slot;
+
+    snprintfz(cases[n - 1].slot_word, sizeof(cases[n - 1].slot_word) - 1,
+              PLUGINSD_KEYWORD_SLOT ":%zu", max_slot + 1);
+    cases[n - 1].expected = 0;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/plugins.d/pluginsd_parser.c:1557]: Both addressed. (1) Confirmed snprintfz() takes the full buffer size (it does vsnprintf(dst, n, ...) then forces dst[n-1]=0), so sizeof(buf)-1 just under-used the buffer by a byte -- changed all three calls to pass sizeof(buf). (2) Added an early guard: the helper returns failure if max_slot < 1, since the max_slot-1 boundary case would otherwise underflow. Real callers only pass nonzero compile-time caps, but the guard makes the helper safe against misuse.

File: src/plugins.d/pluginsd_internals.h [hunk 4]
```diff
@@ -369,13 +369,34 @@ static inline RRDSET *pluginsd_find_chart(RRDHOST *host, const char *chart, cons
...
+        unsigned long long parsed_slot = str2ull_encoded(&id[5]);
+        if(unlikely(parsed_slot > max_slot)) {
+            nd_log_limit_static_global_var(erl_slot, 1, 0);
+            nd_log_limit(&erl_slot, NDLS_COLLECTORS, NDLP_WARNING,
+                         "PLUGINSD: ignoring invalid SLOT value '%s' above the supported maximum %zu",
+                         &id[5], max_slot);
+            // slot 0 means: the SLOT word was present (so the caller still advances its
+            // word index), but the value is unusable as a cache index. Downstream cache
+            // paths treat slot < 1 as "uncached" and fall back to lookup by id.
+            slot = 0;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/plugins.d/pluginsd_internals.h:398]: Good catch on the imprecise wording. Reworded the comment to avoid implying a universal id-fallback. For slot 0 specifically the dimension NULL-error path you mention is not reached: pluginsd_rrddim_put_to_slot() sets dims_with_slots=false for slot < 1, so the chart switches to the non-slotted (by-position) caching path rather than the slotted branch in pluginsd_acquire_dimension() that returns NULL. The comment now states: chart lookups fall back to find-by-id, and dimension caching switches to the non-slotted path.

**Review by ilyam8 — APPROVED** [review]
